File: 2023/day_3/rust/src/bin/part_one.rs
fn main() {
    let lines: Vec<&str> = include_str!("../../../input.txt").lines().collect();
    let mut numbers: Vec<u32> = Vec::new();

    for (line_idx, line) in lines.iter().enumerate() {
        let mut look_after = 0;

        for (char_idx, char) in line.char_indices() {
            if char.is_numeric() && char_idx >= look_after {
                let mut number = String::new();

                look_after = char_idx;
                while let Some(other_char) = line.chars().nth(look_after) {
                    if other_char.is_numeric() {
                        number.push(other_char);
                        look_after += 1;
                    } else {
                        break;
                    }
                }

                let l_r_chars = [
                    char_idx
                        .checked_sub(1)
                        .and_then(|idx| line.chars().nth(idx)),
                    line.chars().nth(look_after),
                ];

                let adjacent_lines = [
                    line_idx.checked_sub(1).and_then(|idx| lines.get(idx)),
                    lines.get(line_idx + 1),
                ];

                let indices = char_idx.saturating_sub(1)..look_after + 1;
                let adjacent_chars = adjacent_lines
                    .iter()
                    .flatten()
                    .flat_map(|line| indices.clone().flat_map(|i| line.chars().nth(i)));

                for char in adjacent_chars.chain(l_r_chars.into_iter().flatten()) {
                    if !char.is_numeric() && char != '.' {
                        numbers.push(number.parse().unwrap());
                    }
                }
            }
        }
    }

    println!("sum: {:?}", numbers.iter().sum::<u32>());
}

File: 2023/day_1/c++/part_one.cpp
#include <iostream>
#include <vector>
#include <fstream>

const char numbers[] = {'1', '2', '3', '4', '5', '6', '7', '8', '9', '0'};

int main()
{
    std::ifstream input("../input.txt");
    std::vector<int> current_line;
    int sum = 0;

    char ch;
    while (input.get(ch))
    {
        if (ch == '\n')
        {
            sum += 10 * current_line[0] + current_line[current_line.size() - 1];
            current_line.clear();
            continue;
        }

        for (char number : numbers)
        {
            if (ch == number)
            {
                current_line.push_back(ch - '0');
                break;
            }
        }
    }

    std::cout << sum << std::endl;

    input.close();
    return 0;
}
